src/protocolsupport/protocol/packet/middleimpl/clientbound/play/v_pe/CraftingData.java

```diff
@@ -0,0 +1,143 @@
+import protocolsupport.protocol.serializer.MiscSerializer;
+import protocolsupport.protocol.serializer.VarNumberSerializer;
+import protocolsupport.protocol.typeremapper.pe.PEItems;
+import protocolsupport.protocol.typeremapper.pe.PEPacketIDs;
+import protocolsupport.protocol.utils.i18n.I18NData;
+import protocolsupport.protocol.utils.types.NetworkItemStack;
+import protocolsupport.utils.recyclable.RecyclableCollection;
+import protocolsupport.utils.recyclable.RecyclableSingletonList;
+
+import java.util.LinkedList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.UUID;
+import java.util.function.Consumer;
+
+public class CraftingData extends MiddleDeclareRecipes {
+	protected int recipesWritten;
+
+	public CraftingData(ConnectionImpl connection) {
+		super(connection);
+	}
+
+	void expandRecipe(Ingredient[] ingredients, Consumer<List<NetworkItemStack>> consumer) {
+		expandRecipe(new LinkedList<>(Arrays.asList(ingredients)), new LinkedList<>(), consumer);
+	}
+
+	void expandRecipe(LinkedList<Ingredient> ingredients, List<NetworkItemStack> items, Consumer<List<NetworkItemStack>> consumer) {
+		LinkedList<Ingredient> tail = new LinkedList<>(ingredients);
+		Ingredient first = tail.removeFirst();
+		NetworkItemStack[] possibleStacks = first.getPossibleStacks();
+		if (possibleStacks.length == 0) { //empty
+			possibleStacks = new NetworkItemStack[]{null};
+		}
+		if (possibleStacks.length > 2) { //'large' ingredient set
+			NetworkItemStack firstType = ItemStackSerializer.remapItemToClient(connection.getVersion(), I18NData.DEFAULT_LOCALE, possibleStacks[0].cloneItemStack());
+			boolean allSameType = true;
+			//see if all ingredients are variants of the same type
+			for (NetworkItemStack ing : possibleStacks) {
+				NetworkItemStack thisType = ItemStackSerializer.remapItemToClient(connection.getVersion(), I18NData.DEFAULT_LOCALE, ing.cloneItemStack());
+				if(thisType.getTypeId() != firstType.getTypeId() || thisType.getAmount() != 1) {
+					allSameType = false;
+					break;
+				}
+			}
+			if (allSameType) {
+				NetworkItemStack wildStack = new NetworkItemStack();
+				wildStack.setTypeId(firstType.getTypeId());
+				wildStack.setLegacyData(-1); //wildcard stack
+				wildStack.setAmount(1);
+				possibleStacks = new NetworkItemStack[]{wildStack};
+			}
+		}
+		for (NetworkItemStack item : possibleStacks) {
+			LinkedList<NetworkItemStack> newItems = new LinkedList<>(items);
+			newItems.add(item);
+			if (tail.isEmpty()) {
+				consumer.accept(newItems); //final step
+			} else {
+				//recursively call with each new level of permutation
+				expandRecipe(tail, newItems, consumer);
+			}
+		}
+	}
+
+	@Override
+	public RecyclableCollection<ClientBoundPacketData> toData() {
+		recipesWritten = 0;
+		ByteBuf buffer = Unpooled.buffer();
+		for (Recipe recipe : recipes) {
+			if (recipe instanceof ShapelessRecipe) {
+				ShapelessRecipe shapeless = (ShapelessRecipe)recipe;
```

Comment:
Recipe)\*space\*

Reply:
@7kasper 
Is there some code style rules for the project? It would be nice to have it encoded so we could apply automatic formatting.

Reply:
Its fairly common styling, but with a bit Shevchik flavour..
The default eclipse Ctrl+Alt+F encodes it right most of the time.